pkg/operator/controller/ingress_status.go

```diff
@@ -7,12 +7,145 @@ import (
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{string(c.Status.EndpointPublishingStrategy.Type)}, nil
+}
+
+func indexByProvisioned(obj interface{}) ([]string, error) {
+	service, ok := obj.(*corev1.Service)
+	if !ok {
+		return []string{}, nil
+	}
+	return controlledBy(service, isProvisioned), nil
+}
+
+func indexByPending(obj interface{}) ([]string, error) {
+	service, ok := obj.(*corev1.Service)
+	if !ok {
+		return []string{}, nil
+	}
+	return controlledBy(service, isPending), nil
+}
+
+func indexContains(indexer toolscache.Indexer, name, value string) bool {
+	objects, err := indexer.ByIndex(name, value)
+	if err != nil {
+		return false
+	}
+	return len(objects) > 0
+}
+
+var serviceIndexers = toolscache.Indexers{
+	"pending-for":     indexByPending,
+	"provisioned-for": indexByProvisioned,
+}
+var controllerIndexers = toolscache.Indexers{
+	"publishing": indexByPublishing,
+}
+
+func NewIngressStatus(controllers, services toolscache.Indexer) *ingressStatus {
+	controllers.AddIndexers(controllerIndexers)
+	services.AddIndexers(serviceIndexers)
+	return &ingressStatus{
+		controllers: controllers,
+		services:    services,
+	}
+}
+
+func (s *ingressStatus) computeLoadBalancerStatus(ic *operatorv1.IngressController) []operatorv1.OperatorCondition {
+	if ic.Status.EndpointPublishingStrategy.Type != operatorv1.LoadBalancerServiceStrategyType {
+		return []operatorv1.OperatorCondition{
+			{
+				Type:    "LoadBalancerManaged",
+				Status:  operatorv1.ConditionFalse,
+				Reason:  "UnsupportedEndpointPublishingStrategy",
+				Message: fmt.Sprintf("The %s endpoint publishing strategy does not support a load balancer", ic.Status.EndpointPublishingStrategy.Type),
+			},
+		}
+	}
+
+	conditions := []operatorv1.OperatorCondition{}
+
+	conditions = append(conditions, operatorv1.OperatorCondition{
+		Type:    "LoadBalancerManaged",
+		Status:  operatorv1.ConditionTrue,
+		Reason:  "LoadBalancerEndpoingPublishingStrategy",
+		Message: "The LoadBalancer endpoint publishing strategy is specified",
+	})
+
+	pending := indexContains(s.services, "pending-for", ic.Name)
+	provisioned := indexContains(s.services, "provisioned-for", ic.Name)
+
+	switch {
+	case pending:
+		conditions = append(conditions, operatorv1.OperatorCondition{
+			Type:    "LoadBalancerAvailable",
+			Status:  operatorv1.ConditionFalse,
+			Reason:  "LoadBalancerPending",
+			Message: "The LoadBalancer service is pending",
+		})
+	case provisioned:
+		conditions = append(conditions, operatorv1.OperatorCondition{
+			Type:    "LoadBalancerAvailable",
+			Status:  operatorv1.ConditionTrue,
+			Reason:  "LoadBalancerProvisioned",
+			Message: "The LoadBalancer service is provisioned",
+		})
+	default:
+		conditions = append(conditions, operatorv1.OperatorCondition{
+			Type:    "LoadBalancerAvailable",
+			Status:  operatorv1.ConditionFalse,
+			Reason:  "LoadBalancerStatusUnknown",
+			Message: "The LoadBalancer service status is unknown",
```

Comment:
The union of "pending-for" and "provisioned-for" is all LB services that are owned by ingress controllers, right? So if neither index had an entry for the ingress controller, that means the service is non-existent, right? So the status us now unknown, but rather that the LB service does not exist.

Reply:
Yes, good point — I changed the default to `LoadBalancerNotFound` and added a test.